controllers/iperf3/clientpod.go

```diff
@@ -17,32 +17,51 @@ limitations under the License.
 package iperf3
 
 import (
-	"context"
 	"fmt"
 	"strings"
 
-	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
-
 	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 
-	"github.com/xridge/kubestone/pkg/k8s"
-
 	perfv1alpha1 "github.com/xridge/kubestone/api/v1alpha1"
 )
 
 // +kubebuilder:rbac:groups="",resources=pods,verbs=get;list;create;delete
 
-func (r *Reconciler) newClientPod(ctx context.Context, cr *perfv1alpha1.Iperf3, crRef *corev1.ObjectReference) error {
-	iperfCmdLineArgs := fmt.Sprintf("-c %s %s", cr.Name, cr.Spec.ClientConfiguration.CmdLineArgs)
+func clientPodName(cr *perfv1alpha1.Iperf3) string {
+	// Should not match with service name as the pod's
+	// hostname is set to it's name. If the two matches
+	// the destination ip will resolve to 127.0.0.1 and
+	// the server will be unreachable.
+	return serverServiceName(cr) + "-client"
+}
+
+// NewClientPod creates an Iperf3 Client Pod (targetting the
+// Server Deployment via the Server Service) from the provided
+// IPerf3 Benchmark Definition.
+func NewClientPod(cr *perfv1alpha1.Iperf3) *corev1.Pod {
+	serverAddress := serverServiceName(cr)
+	iperfCmdLineArgs := fmt.Sprintf("--client %s --port %d ",
+		serverAddress, Iperf3ServerPort)
+
+	if cr.Spec.UDP {
+		iperfCmdLineArgs += "--udp "
+	}
+
+	iperfCmdLineArgs += cr.Spec.ClientConfiguration.CmdLineArgs
 	pod := corev1.Pod{
 		ObjectMeta: metav1.ObjectMeta{
 			Labels:    cr.Spec.ClientConfiguration.PodLabels,
-			Name:      cr.Name + "-client",
+			Name:      clientPodName(cr),
 			Namespace: cr.Namespace,
 		},
 		Spec: corev1.PodSpec{
 			RestartPolicy: corev1.RestartPolicyNever,
+			ImagePullSecrets: []corev1.LocalObjectReference{
+				{
+					Name: cr.Spec.Image.PullSecret,
+				},
+			},
```

Comment:
Be careful as strings.Fields(iperfCmdLineArgs) in line 71 does not handle shell quotation.

Reply:
this is resolved